peers/app_request_network.go

```diff
@@ -103,83 +100,99 @@ func NewNetwork(
+		// Get the list of destination chains that this source may relay to
+		var destinationBlockchainIDs []ids.ID
+		if supportedDsts := sourceBlockchain.GetSupportedDestinations(); supportedDsts.Len() > 0 {
+			destinationBlockchainIDs = supportedDsts.List()
+		} else {
+			for _, dst := range cfg.DestinationBlockchains {
+				destinationBlockchainIDs = append(destinationBlockchainIDs, dst.GetBlockchainID())
+			}
 		}
-	}
-	if len(beaconIPs) == 0 {
-		logger.Error(
-			"Failed to find any peers to connect to",
-			zap.Error(err),
-		)
-		return nil, nil, err
-	}
-	if len(beaconIPs) < numInitialTestPeers {
-		logger.Warn(
-			"Failed to find a full set of peers to connect to on startup",
-			zap.Int("connectedPeers", len(beaconIPs)),
-			zap.Int("expectedConnectedPeers", numInitialTestPeers),
-		)
-	}
 
-	for i, beaconIDStr := range beaconIDs {
-		beaconID, err := ids.NodeIDFromString(beaconIDStr)
-		if err != nil {
-			logger.Error(
-				"Failed to parse beaconID",
-				zap.String("beaconID", beaconIDStr),
-				zap.Error(err),
+		var connectedValidators *ConnectedCanonicalValidators
+		if sourceBlockchain.GetBlockchainID() == constants.PrimaryNetworkID {
+			// For the primary network, connect to the validators of the destination chains
+			for _, destinationBlockchainID := range destinationBlockchainIDs {
+				subnetID, err := pChainClient.ValidatedBy(context.Background(), destinationBlockchainID)
+				if err != nil {
+					logger.Error(
+						"Failed to get subnet ID",
+						zap.Error(err),
+					)
+					return nil, nil, err
+				}
+				connectedValidators, err = arNetwork.ConnectToCanonicalValidators(
+					validators.NewCanonicalValidatorClient(logger, pChainClient),
+					subnetID,
+				)
+				if err != nil {
+					logger.Error(
+						"Failed to connect to canonical validators",
+						zap.Error(err),
+					)
+					return nil, nil, err
+				}
+			}
+		} else {
+			// Otherwise, connect to the validators of the source subnet
+			connectedValidators, err = arNetwork.ConnectToCanonicalValidators(
+				validators.NewCanonicalValidatorClient(logger, pChainClient),
+				sourceBlockchain.GetSubnetID(),
 			)
-			return nil, nil, err
+			if err != nil {
+				logger.Error(
+					"Failed to connect to canonical validators",
+					zap.Error(err),
+				)
+				return nil, nil, err
+			}
 		}
 
-		beaconIPStr := beaconIPs[i]
-		ipPort, err := ips.ToIPPort(beaconIPStr)
-		if err != nil {
-			logger.Error(
-				"Failed to parse beaconIP",
-				zap.String("beaconIP", beaconIPStr),
-				zap.Error(err),
-			)
-			return nil, nil, err
+		for _, destinationBlockchainID := range destinationBlockchainIDs {
+			quorum, err := cfg.GetWarpQuorum(destinationBlockchainID)
+			if err != nil {
+				logger.Error(
+					"Failed to get warp quorum from config",
+					zap.Error(err),
+				)
+				return nil, nil, err
+			}
+			if !utils.CheckStakeWeightExceedsThreshold(
+				big.NewInt(0).SetUint64(connectedValidators.ConnectedWeight),
```

Comment:
In the case that the source network is the C-Chain (or generally a chain in the primary network), we assigned `connectedValidators` multiple times in a nested for loop above, but I think only check the quorom here against the final assignment value.

Reply:
This is fixed now.